addons/stock/models/stock_quant.py

```diff
@@ -766,7 +766,7 @@ def _gather(self, product_id, location_id, lot_id=None, package_id=None, owner_i
         removal_strategy = self._get_removal_strategy(product_id, location_id)
         domain = self._get_gather_domain(product_id, location_id, lot_id, package_id, owner_id, strict)
         domain, order = self._get_removal_strategy_domain_order(domain, removal_strategy, qty)
-        if self:
+        if self.ids:
```

Comment:
It's already the definition in BaseModel 
https://github.com/odoo/odoo/blob/300e024097b186932215a349f01db6b6d66cbbbc/odoo/models.py#L6383-L6385

Reply:
`self.ids` != `self._ids`.
When invoked on a new record (e.g. trying to create a new `stock.quant` in this case):
- `self.ids` will be an empty list.
- `self._ids` will be a Tuple containing the `NewId` of the to-be-created record instead, meaning that its boolean evaluation will be `True`.